### ISSUE TITLE
Wrong expected count for CSI controller and node - various volume issues

### ISSUE THREAD

**Issue by ppacher** [hashicorp/nomad#12142]
### Nomad version

Nomad v1.2.6 (a6c6b475db5073e33885377b4a5c733e1161020c)

### Operating system and Environment details

Nomad Cluster with 5 Servers and 2 additional clients spread across two "datacenters".

### Issue

I'm pretty sure I'm facing different issues here but I'm not sure where to draw the line between each of them so I decided to create one for all of them and separate them once I have more information:

- I started having placement issue with my mongodb job because of `CSI volume mongo-data has exhausted its available writer claims` although there are no other allocations running that require the volume (confirmed via `nomad volume status`). Since I had similar issue with volumes not being detached on reschedule I decided to restart the nomad servers/clients once after each other. 
- Now one of my CSI driver plugins (hetzner-cloud, monolith) reports an expected count of 0 (for nodes and controllers) and an actual, healthy count of 4 (actually expected would be 5). 
- In the UI, the health of some modules is reported as "Node Only". Once I run `nomad volume status <name>` on the respective volume it's health "state" switches from "Node Only" to "Healthy (4/0)" (same as reported on the cli).
- Once of my volumes using the faulty provider is stuck in `CSI volume postgres-data has exhausted its available writer claims and is claimed by a garbage collected allocation <alloc-id> waiting for claim to be released` for some time now. Manually running `nomad system gc` doesn't help here.
- 

### Reproduction steps

I'm not sure still trying to investigate the current state of the cluster. Will update here with more information as I find them.

#### Expected Result

CSI Volumes to work as expected.

#### Actual Result

Expected count of 0 for CSI plugin controller and nodes. Attached volumes work but unused volumes are blocked. One volume stuck in "garbage collection" and "exhausted writer claims" for volumes that are not used by any allocation.

### Nomad Server logs

From the current leader (also a client):
```
1+0100 [ERROR] client.rpc: error performing RPC to server: error="rpc error: rpc error: could not detach from node: node detach volume: CSI.NodeDetachVolume: plugin csi.hetzner.cloud for type csi-node not found" rpc=CSIVolume.Unpublish server=10.80.0.1:4647
Feb 28 09:23:14 srv1 nomad[426]:     2022-02-28T09:23:14.461+0100 [ERROR] client.rpc: error performing RPC to server which is not safe to automatically retry: error="rpc error: rpc error: could not detach from node: node detach volume: CSI.NodeDetachVolume: plugin csi.hetzner.cloud for type csi-node not found" rpc=CSIVolume.Unpublish server=10.80.0.1:4647
Feb 28 09:23:14 srv1 nomad[426]:   | hook "csi_hook" failed: 1 error occurred:
Feb 28 09:23:14 srv1 nomad[426]:   |         * rpc error: rpc error: could not detach from node: node detach volume: CSI.NodeDetachVolume: plugin csi.hetzner.cloud for type csi-node not found
Feb 28 09:24:06 srv1 nomad[426]:     2022-02-28T09:24:06.850+0100 [INFO]  client.plugin: shutting down plugin manager: plugin-type=csi
Feb 28 09:24:06 srv1 nomad[426]:     2022-02-28T09:24:06.850+0100 [INFO]  client.plugin: plugin manager finished: plugin-type=csi
Feb 28 09:24:07 srv1 nomad[37253]:     2022-02-28T09:24:07.509+0100 [INFO]  client.plugin: starting plugin manager: plugin-type=csi
Feb 28 09:24:08 srv1 nomad[37253]:     2022-02-28T09:24:08.510+0100 [ERROR] client.csi.hetzhner.cloud: failed to setup instance manager client: error="failed to open grpc connection to addr: /var/nomad/client/csi/monolith/csi.hetzhner.cloud/csi.sock, err: context deadline exceeded"
Feb 28 09:24:09 srv1 nomad[37253]:     2022-02-28T09:24:09.512+0100 [ERROR] client.csi.hetzhner.cloud: failed to setup instance manager client: error="failed to open grpc connection to addr: /var/nomad/client/csi/monolith/csi.hetzhner.cloud/csi.sock, err: context deadline exceeded"
Feb 28 09:24:49 srv1 nomad[37253]:     2022-02-28T09:24:49.400+0100 [INFO]  client.alloc_runner.task_runner.task_hook.logmon.nomad: opening fifo: alloc_id=3fc05dc8-949a-177f-a545-7993695a909e task=hcloud-csi @module=logmon path=/var/nomad/alloc/3fc05dc8-949a-177f-a545-7993695a909e/alloc/logs/.hcloud-csi.stdout.fifo timestamp="2022-02-28T09:24:49.400+0100"
Feb 28 09:24:49 srv1 nomad[37253]:     2022-02-28T09:24:49.401+0100 [INFO]  client.alloc_runner.task_runner.task_hook.logmon.nomad: opening fifo: alloc_id=3fc05dc8-949a-177f-a545-7993695a909e task=hcloud-csi @module=logmon path=/var/nomad/alloc/3fc05dc8-949a-177f-a545-7993695a909e/alloc/logs/.hcloud-csi.stderr.fifo timestamp="2022-02-28T09:24:49.401+0100"
Feb 28 09:24:49 srv1 nomad[37253]:     2022-02-28T09:24:49.729+0100 [INFO]  agent: (runner) rendered "(dynamic)" => "/var/nomad/alloc/3fc05dc8-949a-177f-a545-7993695a909e/hcloud-csi/secrets/env"
Feb 28 09:37:37 srv1 nomad[37253]:   |         * could not detach from node: rpc error: node detach volume: CSI.NodeDetachVolume: plugin csi.hetzner.cloud for type csi-node not found
Feb 28 09:37:37 srv1 nomad[37253]:   |         * could not detach from node: rpc error: node detach volume: CSI.NodeDetachVolume: plugin csi.hetzner.cloud for type csi-node not found
Feb 28 09:37:37 srv1 nomad[37253]:   |         * could not detach from node: rpc error: node detach volume: CSI.NodeDetachVolume: plugin csi.hetzner.cloud for type csi-node not found
Feb 28 09:37:37 srv1 nomad[37253]:   |         * could not detach from node: rpc error: node detach volume: CSI.NodeDetachVolume: plugin csi.hetzner.cloud for type csi-node not found
Feb 28 09:37:37 srv1 nomad[37253]:   |         * could not detach from node: rpc error: node detach volume: CSI.NodeDetachVolume: plugin csi.hetzner.cloud for type csi-node not found
```

**Comment by tgross** [comment]
Hi @ppacher! It looks like you've run into a couple of different known issues with the CSI beta. These have been patched and are expected to ship in Nomad 1.3.0 (with backports to 1.2.x and 1.1.x): https://github.com/hashicorp/nomad/issues/10927 https://github.com/hashicorp/nomad/issues/10052 https://github.com/hashicorp/nomad/issues/10073.

I'm going to close this issue as a duplicate, but in the meantime I can try to help you recover this situation as best we can. You may be able to get the servers to "detach" the unused volume via `nomad volume detach -force`.

**Comment by ppacher** [comment]
Hi @tgross , thanks for the fast reply and for pointing me to the existing issues. I should have done a more throughout search I guess :)

I've been able to detach the unused volumes but Nomad still reports an expected count of 0 for one CSI plugin. Though, it seems like it's happily attaching volumes now so I guess that's fine. Looking forward to v1.3.0 of this awesome project!
